compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPULayoutAnalysisAndDistribution.cpp

```diff
@@ -262,8 +274,74 @@ void setMMALayout(Value aMatrix, Value bMatrix, Value cMatrix,
   setLayout(cMatrix, MMAMatrixType::CMatrix, "cMatrix");
 }
 
+void propagateLayoutToReductions(vector::MultiDimReductionOp reductionOp,
+                                 DenseMap<Value, Layout> &layoutMap) {
+  Value src = reductionOp.getSource();
+  if (!layoutMap.count(src)) return;
+  // Determine layout of result (after reduction)
+  Value result = reductionOp.getResult();
+  Layout resultLayout(layoutMap.at(src));
+  auto reductionDims = reductionOp.getReductionDims().getAsRange<IntegerAttr>();
+  for (auto ia : reductionDims) {
+    resultLayout.reduceDim(ia.getInt());
+  }
+  layoutMap.try_emplace(result, resultLayout);
+  // Propagate result layout to accumulator
+  Value acc = reductionOp.getAcc();
+  layoutMap.try_emplace(acc, resultLayout);
+  layoutMap.at(result).debugPrint("reduced");
+}
+
+void propagateLayoutToBroadcasts(vector::BroadcastOp broadcastOp,
+                                 DenseMap<Value, Layout> &layoutMap) {
+  Value src = broadcastOp.getSource();
+  if (!layoutMap.count(src)) return;
+  // For now, only compute for broadcasts after reductions
+  auto reductionOp = src.getDefiningOp<vector::MultiDimReductionOp>();
+  if (!reductionOp) return;
+  Value reductionSrc = reductionOp.getSource();
+  if (!layoutMap.count(reductionSrc)) return;
+  // For now, require that result shape and source shape are the same
+  // Unless there is a transpose that changes the result shape to the source
+  // shape
+  Value result = broadcastOp.getResult();
+  ArrayRef<int64_t> resultShape =
+      result.getType().cast<ShapedType>().getShape();
+  ArrayRef<int64_t> srcShape =
+      reductionSrc.getType().cast<ShapedType>().getShape();
+  bool success = srcShape == resultShape;
+  if (!success) {
+    for (Operation *user : result.getUsers()) {
+      if (auto transposeOp = dyn_cast<vector::TransposeOp>(user)) {
+        Value transposedResult = transposeOp.getResult();
+        ArrayRef<int64_t> transposedShape =
+            transposedResult.getType().cast<ShapedType>().getShape();
+        if (transposedShape == srcShape) {
+          // We will lower reduction + broadcast + transpose at once
+          // So assign all their results the same layout
+          layoutMap.try_emplace(transposedResult, layoutMap.at(reductionSrc));
+          layoutMap.at(transposedResult).debugPrint("transposed");
+          success = true;
+          break;
+        }
+      }
+    }
+  }
+  if (success) {
+    // Sets layout for either reduction + broadcast (if right shape)
+    // or reduction + broadcast + transpose
+    layoutMap.try_emplace(result, layoutMap.at(reductionSrc));
+    layoutMap.at(result).debugPrint("broadcasted");
+  }
+}
```

Comment:
So just to summarize, these two functions add layouts for
1. Reductions result and accumulator
2. Broadcast result (if broadcast shape = reduction source shape), Broadcast and Transpose result (if transpose shape = reduction source shape).

Now I think the first question is why have a layout for the reductions? I would agree that if we are lowering reduction + broadcast + transpose together, then it might not make sense.

But here is the IR of the innermost loop for flash attention: https://gist.github.com/harsh-nod/d0610bfbbca89bfba5012a7d9165c4b8

Specifically, you look at this
```
%20 = vector.multi_reduction <maxf>, %18, %19 [1] : vector<32x128xf32> to vector<32xf32>
%21 = vector.broadcast %20 : vector<32xf32> to vector<128x32xf32>
%22 = vector.transpose %21, [1, 0] : vector<128x32xf32> to vector<32x128xf32>
%23 = arith.subf %18, %22 : vector<32x128xf32>
%24 = math.exp %23 : vector<32x128xf32>
%25 = vector.transfer_read %arg10[%c0, %arg7], %cst_2 {in_bounds = [true]} : tensor<1x128xf32>, vector<32xf32>
%26 = arith.subf %19, %20 : vector<32xf32>
%27 = math.exp %26 : vector<32xf32>
%28 = arith.mulf %27, %25 : vector<32xf32>
```
Here you can see we are using the reduction result in a subtraction
```
%26 = arith.subf %19, %20 : vector<32xf32>
```
Here unless we compute the layout of the reduction result, it is not clear what layout to use for `%19` and `%26`. So I think we need to compute the layout of the reduction values. 

Now when we have a broadcast + transpose, then I do agree that we don't need to set the layout for the broadcast (because like you said we can't materialize this layout for the broadcast) so I can modify the code to only set it for the transpose in that case. 

Does that make sense? Essentially what I am saying is that we can't skip the layout for intermediate values because they might have other uses with the caveat being the broadcast result when the broadcast and transpose are used together.

Reply:
I see, I was hoping we would convert all the reductions into a reduce_broadcast where the dim reduced is broadcasted in place. Sounds like this is not possible then. 
Handling the cases where not all the lanes have a value mapped to the is going to be difficult I think. You won't be able to know how lanes are there on laneId.x which I'm guessing will break some of the logic. Emitting stores will also require emitting predicate or control flow to only enable some lanes to store. 
It would be good to figure out how this is going to be handled.

Reply:
Okay I think I understand better now. So seems like they key here is to make sure all lanes are participating otherwise the logic gets complicated (doable but still complicated). And the key to have all lanes participating is to stick with the original 2d layout. 

If that's the case, then I believe it is possible to have all the reductions be expressed as reduce_broadcast. To do this, we have to change all the uses of the reduction value to operate instead on the broadcasted value. For example, going back to my original example

```
%20 = vector.multi_reduction <maxf>, %18, %19 [1] : vector<32x128xf32> to vector<32xf32>
%21 = vector.broadcast %20 : vector<32xf32> to vector<128x32xf32>
%22 = vector.transpose %21, [1, 0] : vector<128x32xf32> to vector<32x128xf32>
%23 = arith.subf %18, %22 : vector<32x128xf32>
%24 = math.exp %23 : vector<32x128xf32>
%25 = vector.transfer_read %arg10[%c0, %arg7], %cst_2 {in_bounds = [true]} : tensor<1x128xf32>, vector<32xf32>
%26 = arith.subf %19, %20 : vector<32xf32>
%27 = math.exp %26 : vector<32xf32>
%28 = arith.mulf %27, %25 : vector<32xf32>
```
We would essentially change the last 3 ops to operate on the broadcasted value instead of the reduced value. So I can skip setting the layout for the intermediate results and will modify the code.